dropshot/src/server.rs

```diff
@@ -543,18 +543,55 @@ impl<C: ServerContext> Service<&TlsConn> for ServerConnectionHandler<C> {
     }
 }
 
-/**
- * A running Dropshot HTTP server.
- */
-pub struct HttpServer<C: ServerContext> {
+/// A running Dropshot HTTP server.
+///
+/// The generic traits represent the following:
+/// - C: Caller-supplied server context
+/// - Closer: Identifies if the server is directly closeable.
+/// - Joiner: Identifies if the server is directly joinable.
+pub struct HttpServer<C: ServerContext, Closer = Closeable, Joiner = Joinable> {
     probe_registration: ProbeRegistration,
     app_state: Arc<DropshotState<C>>,
     local_addr: SocketAddr,
-    join_handle: Option<tokio::task::JoinHandle<Result<(), hyper::Error>>>,
+    closer: Closer,
+    joiner: Joiner,
+}
+
+/// Represents the ability to close a running HTTP server.
+///
+/// Returned from [HttpServer::get_closer].
+pub struct Closeable {
```

Comment:
```suggestion
pub struct CloseHandle {
```

I find "Closeable" and "Joinable", both described as "abilities", a bit confusing.  Those sound like they would be traits.  I've suggested "CloseHandle" and "ShutdownWaitHandle" instead and suggested some doc changes.  Let me know what you think.

Reply:
Updated